odoo/addons/base/tests/test_sql.py

```diff
@@ -158,3 +175,18 @@ def test_complex_sql(self):
             repr(sql),
             """SQL('SELECT "id" FROM "table" WHERE "table"."foo"=%s AND "table"."bar"=%s', 1, 2)"""
         )
+
+
+class TestSqlTools(TransactionCase):
+
+    def test_add_constraint(self):
+        definition = "CHECK (name !~ '%')"
```

Comment:
Shouldn't you use some SQL expression that can be passed directly to `cr.execute()`?
If that's the case, we must require escaping, since `cr.execute()` now requires it.

Reply:
The function `add_constraint` escapes the constraint as needed. The last argument is the condition definition as a string.

Reply:
I added the same behaviour to `add_index` so that it handles definitions containing '%'.

Reply:
My actual question is: is it a good idea to make this confusing for developers?
The rule of thumb is: always escape `%` _except for constraint and index definitions_.
Isn't that double standard a source of confusion?

Reply:
I don't think it is a double-standard: the rule is always escape when using `SQL` wrapper and these 2 functions accept a str as input.